src/gurt/shm_alloc.c

```diff
@@ -0,0 +1,296 @@
+			DS_ERROR(errno, "shm_mutex_init() failed");
+			goto err_unmap;
+		}
+	}
+	if (shm_mutex_init(&(d_shm_head->ht_lock)) != 0) {
+		DS_ERROR(errno, "shm_mutex_init() failed");
+		goto err_unmap;
+	}
+
+	pid_shm_creator         = pid;
+	d_shm_head->off_ht_head = INVALID_OFFSET;
+
+	atomic_store_relaxed(&(d_shm_head->ref_count), 1);
+	atomic_store_relaxed(&(d_shm_head->large_mem_count), 0);
+	d_shm_head->size           = shm_size;
+	d_shm_head->shm_pool_size  = shm_pool_size;
+	d_shm_head->magic          = DSM_MAGIC;
+	/* initialization is finished now. */
+	return 0;
+
+err_unmap:
+	d_shm_head = NULL;
+	munmap(shm_addr, shm_size);
+
+err:
+	close(shm_ht_fd);
+	return errno;
+}
+
+static void *
+shm_alloc_comm(size_t align, size_t size)
+{
+	int      idx_allocator;
+	void    *buf;
+	int      tid;
+	uint32_t hash;
+	uint64_t oldref;
+
+	if (idx_small < 0) {
+		tid  = syscall(SYS_gettid);
+		hash = d_hash_string_u32((const char *)&tid, sizeof(int));
+		/* choose a memory allocator based on tid */
+		idx_small = hash % N_SHM_POOL;
+	}
+	idx_allocator = idx_small;
+	if (size >= LARGE_MEM) {
+		oldref = atomic_fetch_add_relaxed(&(d_shm_head->large_mem_count), 1);
+		/* pick the allocator for large memory request with round-robin */
+		idx_allocator = oldref % N_SHM_POOL;
+	}
+	shm_mutex_lock(&(d_shm_head->mem_lock[idx_allocator]), NULL);
+	if (align == 0)
+		buf = tlsf_malloc(d_shm_head->tlsf[idx_allocator], size);
+	else
+		buf = tlsf_memalign(d_shm_head->tlsf[idx_allocator], align, size);
+	shm_mutex_unlock(&(d_shm_head->mem_lock[idx_allocator]));
+
+	return buf;
+}
+
+void *
+shm_alloc(size_t size)
+{
+	return shm_alloc_comm(0, size);
+}
+
+void *
+shm_memalign(size_t align, size_t size)
+{
+	return shm_alloc_comm(align, size);
+}
+
+void
+shm_free(void *ptr)
+{
+	uint32_t idx_allocator;
+
+	/* compare with the lower bound address of shared memory pool */
+	if (ptr < d_shm_head->tlsf[0]) {
+		DS_ERROR(EINVAL, "Out of range memory pointer for shm_free()\n");
+		return;
+	}
+
+	idx_allocator = ((uint64_t)ptr - (uint64_t)d_shm_head->tlsf[0]) / d_shm_head->shm_pool_size;
+	/* compare with the upper bound address of shared memory pool */
+	if (idx_allocator >= N_SHM_POOL) {
+		DS_ERROR(EINVAL, "Out of range memory pointer for shm_free()\n");
+		return;
+	}
+
+	shm_mutex_lock(&(d_shm_head->mem_lock[idx_allocator]), NULL);
+	tlsf_free(d_shm_head->tlsf[idx_allocator], ptr);
+	shm_mutex_unlock(&(d_shm_head->mem_lock[idx_allocator]));
+}
+
+void
+shm_destroy(void)
+{
+	char daos_shm_file_name[128];
+
+	sprintf(daos_shm_file_name, "/dev/shm/%s_%d", daos_shm_name, getuid());
+	unlink(daos_shm_file_name);
+}
+
+bool
+shm_inited(void)
+{
+	if (d_shm_head == NULL)
+		return false;
+	if (d_shm_head->magic != DSM_MAGIC)
+		return false;
+
+	return true;
+}
+
+void
+shm_fini(void)
+{
+	if (!shm_inited())
+		return;
+
+	atomic_fetch_add_relaxed(&(d_shm_head->ref_count), -1);
+	if (pid != pid_shm_creator)
+		munmap(d_shm_head, d_shm_head->size);
```

Comment:
From my understanding, all the process should `unmap` and then `close` the file.
The shared mmaped file content should be keept until the `unlink()` and `shm_unlink()` will be called.
Thus, not unmapping for the process creating the shared memory file seems to be useless.
I am also concerned that it could be seen as memoy leak by valgrind or other memory checker tools.

Reply:
Considering the cache in kernel space, I thought we may want to keep our cache persistent too. Otherwise, shared memory needs to be initialized again and again. Ideally, the space for caching would be freed after the content expires. We need our own shared memory allocator to dynamically expand/shrink shared memory region. It would be a long way to get there.
Yes. I did have a little concern about whether valgrind can detect the memory leak in shared memory usage here. I could play with valgrind to find out with simple test.

Reply:
Not sure that it will change something for the kernel cache to keep a dangling pointer when the process will die.
From my understanding what will make the difference is to unlink the file.
@mchaarawi do you have an opinion on this ?

Reply:
One extreme case, the caching does not have any benefit at all if a user runs job serially in case we destruct caching once application ends.

Reply:
Maybe, I am missing something, but if you remove the shared file then keeping a dangling pointer will not help if you have unlink the file.  On the other hand, if you do not unlink the file, the cache will still be available even if you do not have a dangling pointer at the end your applications.
However, it is perfectly that I am missing something.

Reply:
shm_destroy() is not called by regular applications. It was called only in the shm_test. The file associated with shared memory will not be unlinked.
I talked to Mohamad recently. He suggested daos_agent will initialize and destruct shared memory region. We will update it later.

Reply:
not sure i quite understood the discussion. the cache needs to be persistent beyond one process lifetime and can be destroyed when the agent is killed.
but as i understand each procecss needs to call shm_open(), mmap(), close(), munmap() on the shared memory region, and only the agent calls shm_unlink().
the agent part can be done later.

Reply:
Thank you!
To be clear, line 209,
"close(shm_ht_fd);"
It only frees the file descriptor.